src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,489 @@
+            queue.faulty_blocks.count() -
+            queue.options.tables_max * lsm_table_content_blocks_max;
+        return faulty_blocks_free;
+    }
+
+    /// Queue a faulty block to request from the cluster and repair.
+    pub fn request_block(queue: *GridRepairQueue, address: u64, checksum: u128) void {
+        assert(!queue.canceling);
+        assert(queue.request_blocks_available() > 0);
+        assert(queue.faulty_tables.count <= queue.options.tables_max);
+
+        if (queue.faulty_blocks.getPtr(address)) |fault| {
+            assert(fault.checksum == checksum);
+            assert(fault.state != .aborting);
+            return;
+        }
+
+        queue.faulty_blocks.putAssumeCapacityNoClobber(address, .{
+            .checksum = checksum,
+            .progress = .block,
+        });
+    }
+
+    pub fn request_table_queued(queue: *GridRepairQueue, address: u64, checksum: u128) bool {
+        assert(!queue.canceling);
+        maybe(queue.request_blocks_available() == 0);
+        assert(queue.faulty_tables.count <= queue.options.tables_max);
+
+        var tables = queue.faulty_tables.peek();
+        while (tables) |queue_table| : (tables = queue_table.next) {
+            if (queue_table.index_address == address) {
+                assert(queue_table.index_checksum == checksum);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    pub fn request_table(
+        queue: *GridRepairQueue,
+        callback: *const fn (*RepairTable, RepairTableResult) void,
+        table: *RepairTable,
+        address: u64,
+        checksum: u128,
+    ) void {
+        assert(!queue.canceling);
+        assert(queue.faulty_tables.count < queue.options.tables_max);
+        assert(!queue.request_table_queued(address, checksum));
+
+        var tables = queue.faulty_tables.peek();
+        while (tables) |queue_table| : (tables = queue_table.next) assert(queue_table != table);
+
+        table.* = .{
+            .callback = callback,
+            .index_address = address,
+            .index_checksum = checksum,
+        };
+        queue.faulty_tables.push(table);
+
+        if (queue.faulty_blocks.getPtr(address)) |fault| {
+            assert(fault.checksum == checksum);
+            assert(fault.progress == .block);
+            assert(fault.state != .aborting);
+
+            fault.progress = .{ .table_index = .{ .table = table } };
+        } else {
+            queue.faulty_blocks.putAssumeCapacityNoClobber(address, .{
+                .checksum = checksum,
+                .progress = .{ .table_index = .{ .table = table } },
+            });
+        }
+    }
+
+    pub fn repair_ready(queue: *const GridRepairQueue, address: u64, checksum: u128) bool {
+        const fault_index = queue.faulty_blocks.getIndex(address) orelse return false;
+        const fault = &queue.faulty_blocks.entries.items(.value)[fault_index];
+        return fault.checksum == checksum and fault.state == .waiting;
```

Comment:
`.waiting` and  `ready` read like antonyms to me.

Reply:
What about `repair_waiting`? repair_in_progress / _scheduled / enqueued (as you suggest elsewhere) sound like they would return true when the block has arrived but is being written.

Reply:
Yeah, that's the simplest&best